applications/OptimizationApplication/python_scripts/controls/material/simp_control.py

```diff
@@ -0,0 +1,199 @@
+        keys = physical_gradient_variable_container_expression_map.keys()
+        if len(keys) != 2:
+            raise RuntimeError(f"Not provided required gradient fields for control \"{self.GetName()}\". Following are the variables:\n\t" + "\n\t".join([k.Name() for k in keys]))
+        if  Kratos.DENSITY not in keys or Kratos.YOUNG_MODULUS not in keys:
+            raise RuntimeError(f"The required gradient for control \"{self.GetName()}\" w.r.t. DENSITY or YOUNG_MODULUS not found. Followings are the variables:\n\t" + "\n\t".join([k.Name() for k in keys]))
+
+        # first calculate the density partial sensitivity of the response function
+        d_j_d_density = physical_gradient_variable_container_expression_map[Kratos.DENSITY]
+
+        # second calculate the young modulus partial sensitivity of the response function
+        d_j_d_youngs = physical_gradient_variable_container_expression_map[Kratos.YOUNG_MODULUS]
+
+        # now calculate the total sensitivities of density w.r.t. phi
+        d_density_d_phi = KratosOA.ControlUtils.SigmoidalProjectionUtils.CalculateForwardProjectionGradient(self.phi, self.materials.GetPhi(), self.materials.GetDensities(), self.beta, 1)
+
+        # now calculate the total sensitivities of young modulus w.r.t. phi
+        d_young_modulus_d_phi = KratosOA.ControlUtils.SigmoidalProjectionUtils.CalculateForwardProjectionGradient(self.phi, self.materials.GetPhi(), self.materials.GetYoungModulus(), self.beta, self.simp_power_fac)
+
+        # now compute response function total sensitivity w.r.t. phi
+        d_j_d_phi = d_j_d_density * d_density_d_phi + d_j_d_youngs * d_young_modulus_d_phi
+
+        return self.filter.FilterIntegratedField(d_j_d_phi)
+
+    def Update(self, control_field: ContainerExpressionTypes) -> bool:
+        if not IsSameContainerExpression(control_field, self.GetEmptyField()):
+            raise RuntimeError(f"Updates for the required element container not found for control \"{self.GetName()}\". [ required model part name: {self.model_part.FullName()}, given model part name: {control_field.GetModelPart().FullName()} ]")
+
+        if Kratos.Expression.Utils.NormL2(self.phi - control_field) > 1e-15:
+            self.phi = self.filter.FilterField(control_field)
+            self.un_buffered_data.SetValue("filtered_phi", self.phi.Clone(), overwrite=True)
+
+            self.__UpdateBeta()
+
+            self.__ApplyDensityAndYoungsModulus()
+
+    def __UpdateBeta(self) -> None:
+        if self.beta_adaptive:
+            step = self.optimization_problem.GetStep()
+            if step % self.beta_update_period == 0 and self.beta_computed_step != step:
+                self.beta_computed_step = step
+                self.beta = min(self.beta * self.beta_increase_frac, self.beta_max)
+                Kratos.Logger.PrintInfo(f"::{self.GetName()}::", f"Increased beta to {self.beta}.")
+
+    def __ApplyDensityAndYoungsModulus(self) -> None:
```

Comment:
Since you are updating the density and you modulus i recommend to name the function UpdateDensityAndYoungModulus